2486 dependency tab on workloads page

opudrovs opened the pull request:
Closes #2486 

- Added `DependsOn` field to the `Kustomization` and `HelmRelease` objects.
- Fixed Automation type definition (switched to using union instead of intersection).
- Updated Kustomization and HelmRelease tests.
- Added a utility function to get `sourceRef`s for automations.
- Added the `useListObjects` hook.
- Added `FluxObjectNode` type for working with DAG graph nodes.
- Added the `DependenciesView` component which displays either dependencies graph or the `No Dependencies` message (added a `shadow` prop to the `Flex` component for this).
- Added the`d3-dag` library.
- Added the `DagGraph` and `DagGraphNode` components to display directed acyclic graphs based on provided list of nodes (thank you, @ozamosi , for the prototype graph code! ⭐  ).
- Added utility functions for getting graph nodes in the current dependencies tree and tests for them.
- Added highlighting the name of the current node in the graph in bold.
- Centered the DAG graph (its default positioning seems to be different from the existing d3-only `DirectedGraph`).
- Added link/edge arrows.

Tested the code with multiple independent dependencies tree (the tree, relevant to the current node is detected correctly), with single and multiple root nodes and with dependencies in the same and across different namespaces:

Single root node:

<img width="1513" alt="Screenshot 2022-09-12 at 12 11 38" src="https://user-images.githubusercontent.com/8824708/189650547-9c921663-4209-404a-bf1a-9916c6d6720b.png">

Multiple root nodes:

<img width="1492" alt="Screenshot 2022-09-12 at 12 11 22" src="https://user-images.githubusercontent.com/8824708/189650538-bd3ce167-32dd-44b2-9961-2f3a2f36ca00.png">

Multiple root nodes across different namespaces:

<img width="1502" alt="Screenshot 2022-09-12 at 14 13 16" src="https://user-images.githubusercontent.com/8824708/189650553-ddb1c23c-15a4-42cb-b5ce-039e219907f7.png">

The No Dependencies message:

<img width="1499" alt="Screenshot 2022-09-12 at 14 55 36" src="https://user-images.githubusercontent.com/8824708/189659500-38e21727-09ed-4eaa-ba3e-fe4e210d26e9.png">

joshri commented on ui/components/DependenciesView.tsx:
```diff
@@ -0,0 +1,141 @@
+
+  return (
+    <RequestStateHandler loading={isLoading} error={error}>
+      {!isLoading && (
```
You might not need this loading condition here - I haven't used the RequestStateHandler component before (maybe I should've) - but it seems like it's handling the loading case. If you still need the loading condition then the RequestStateHandler thing doesn't seem like it's doing it's job and should probably be removed

opudrovs replied:
I thought RequestStateHandler is used mostly for displaying a loading indicator or an error message. Let me check how it works if I remove the inner condition.

opudrovs replied:
Good catch, removed the unneeded inner loading condition, the loading indicator is displayed without it and the content is hidden, while loading data, as expected.

The RequestStateHandler was used in Reconciliation graph, it is where I picked it from.

joshri reviewed (commented):
WOW good stuff in here! I want to combine the graph components if we can, but the bottom line is that it works!

opudrovs commented on ui/components/DagGraph.tsx:
```diff
@@ -0,0 +1,205 @@
+  const nodeHalfHeight = nodeSize.height / 2;
+  const linkStrokeWidth = 5;
+
+  //use d3 to create DAG structure
```
Maybe we could move the shared functionality like zooming/dragging/panning to a hook.

opudrovs replied:
And the API calls to d3 or d3-dag libraries are different in both graphs.

opudrovs replied:
So, there will be constant branching logic "if this graph type then...".

Usually, AFAIK< different chart types are not combines into a single component.

ozamosi replied:
The difference seems to be:
* Shape of arrows (needs conditionals?)
* Component to render objects (can be provided)
* Right now, the the input to `ReconciliationGraph` takes a tree for d3-hierarchy while the input to `DagGraph` takes a flat list for d3-dag.

I think I would drop `d3-hierarchy` entirely, and change the existing graph to generate a flat list of objects again, so that `DagGraph` can render all graphs. 
 
However. I think it's a good idea to do that Another Day™ - this is a lot of new code, so let's wait a couple of releases before we decide if the old code should be migrated to this or not.

opudrovs replied:
OK, sounds good to me. I could try to combine the graphs now but maybe Product decides that they want the new one to look different. And if it requires serious rework of the existing graph component, I think that yes, it's better to do it in a separate ticket.

Another difference between the graphs is offset to shift the graph horizontally to center it. And I somewhat changed the way the lines are drawn int he new graph (int he old graph they are draw between node centers and pass under the nodes what is not needed in general, but it can be combined). The main diff is input data format, yes.

I guess, ReconciliationGraph should be a think wrapper component. All non-reconciliation-specific functionality should be moved to the DirectedGraph anyway even if we don't combine the graphs.

opudrovs replied:
Arrows were not drawn in the original graph, they were added as a separate path in the new one.

opudrovs commented on ui/components/DependenciesView.tsx:
```diff
@@ -0,0 +1,141 @@
+
+  const isLoading = isLoadingData && !graphNodes;
+
+  const shouldShowGraph = !!graphNodes && graphNodes.length > 0;
```
This component will render the Message component if the data finished loading (= graphNodes have been set to a non-null value no matter if there is an error or not, but if there was an error they will be an empty array) and the length of graphNodes is 0. Before the data finished loading the component should not display the No Dependencies message.

opudrovs replied:
Ah, you probably meant not to show the No Dependencies message if there is an error?

opudrovs replied:
So, if the error happens, it can still mean that there are dependencies, just smth. happened while retrieving the objects. We don't know if the object is in a dependencies tree until we load all objects of this kind and build the tree, because the object can be a root node and itself have no dependencies.

So, for now I'll keep displaying the No Dependencies message if there is an error when loading data, probably. TO provide some information to the user. @ozamosi should I display the No Dependencies message if an error happened or should I display nothing? The actual error message will be displayed by the wrapper `RequestStateHandler` component.

ozamosi replied:
This is OK as it is.

opudrovs replied:
OK, good.

opudrovs replied:
Ah, no, checked how RequestStateHandler component works. It will not show any wrapped content if there is an error. It returns earlier with the error message Alert. It's a pity.

opudrovs commented on ui/components/DependenciesView.tsx:
```diff
@@ -0,0 +1,141 @@
+      background-color: #ffffffd9;
+      color: ${(props) => props.theme.colors.neutral30};
+
+      h2 {
```
Like maybe pass a  tag as prop and use span if it's not provided? Or just use isBlock.

opudrovs commented on ui/components/DagGraph.tsx:
```diff
@@ -0,0 +1,205 @@
+                  <g key={index}>
+                    <path
+                      fill="#7a7a7a"
+                      d={`M${l.target.x}, ${l.target.y}
```
The graph state should be determined by props provided to the graph component. Zooming/panning/whatever will still trigger re-renders and node recalculations.

ozamosi commented on ui/components/DagGraphNode.tsx:
```diff
@@ -0,0 +1,127 @@
+  }
+}
+
+const StyledObjectName = styled.span<{ isCurrentNode: boolean }>`
```
> is there a chance

There's always a chance :grin:

Though, now that I've seen it, I kind-of want this feature in the regular graph as well - did we not have something like it once?

> Passed object kind has also changed to FluxObjectNode

Ugh, so it _would_ work, but it would kinda only work by accident as the types are completely independent? That's unfortunate - perhaps that should have further refactoring done to unify the the types before we slam them into the same function.

opudrovs replied:
The old graph itself uses `descendants: any[];` and children used to create `descendants` are of type UnstructuredObject | any[]. Type of rootNode is... I didn't quite get it, there are not many type deifinitions.

The new graph uses FluxObejctNode[] as nodes prop, what is easier.

I could try to combine the graphs or more like replace the old graph with the new one in this PR, but there is a risk that the Dependencies will not make it into this release.

opudrovs replied:
@ozamosi 

>Though, now that I've seen it, I kind-of want this feature in the regular graph as well - did we not have something like it once?

Do you mean the arrows as "this feature" or anything else you want to have in the regular graph too?

ozamosi replied:
> Do you mean the arrows as "this feature" or anything else you want to have in the regular graph too?

Nono, I mean the `isCurrentNode` - which is the thing that changed from GraphNode?

> I could try to combine the graphs or more like replace the old graph with the new one in this PR

It sounds like you're saying they're false friends - they look almost the same, but they're not the same, but they could be made to be the same, but we ought to make the old code look more like the new code. So that sounds like it's out of scope for here.

opudrovs replied:
@ozamosi 

isCurrentNode is very useful for the new graph because the current not is not necessarily the root node so we can see it on any place in the graph and need to be able to tell immediately where the current node is.

But in the reconciliation graph the current node is located as the "second node" under the root node (which is the source). So, you would still like to mark the "second node" as the current node and display its name in bold?

opudrovs replied:
> It sounds like you're saying they're false friends - they look almost the same, but they're not the same, but they could be made to be the same, but we ought to make the old code look more like the new code.

Smth. like that, yeah. 🤣  OK, I will save it for after the next two releases then.

opudrovs replied:
For now combines both node components into `GraphNode`, this was much easier to do than combining the graphs. Now it supports both node data object formats and `isCurrentNode` field for both graphs.

opudrovs commented on ui/components/DependenciesView.tsx:
```diff
@@ -0,0 +1,141 @@
+function UnstyledMessage({ className }: MessageProps) {
+  return (
+    <Flex className={className} wide tall center>
+      <Flex column align shadow>
```
OK, I will look into making the message box a reusable component.

opudrovs replied:
Added a styled MessageFlex component in Flex.ts for message boxes.

ozamosi commented on ui/lib/utils.ts:
```diff
@@ -161,3 +169,122 @@ export const convertImage = (image: string) => {
+  return dependencyNodes.concat(dependentNodes);
+}
+
+// getGraphNodes returns all nodes in the current node's dependency tree, including the current node.
```
Maybe we could move (some of) these functions to e.g. lib/utils/dependencies.ts? The names are a bit generic when they're in the file `utils` (graph, node and neighbour could all mean different things depending on context).

opudrovs replied:
OK, will do it.

opudrovs replied:
Done.

ozamosi commented on ui/lib/utils.ts:
```diff
@@ -161,3 +169,122 @@ export const convertImage = (image: string) => {
+    graphNodes = graphNodes.concat(node);
+  }
+
+  // Remove duplicates from graphNodes.
```
I think if `nodesToExplore` was just `[currentNode]`, `visitedNodes` was initialized with `currentNode.id`, and the early return if `length === 1` was done against `graphNodes` here after the loop instead, then you couldn't get duplicates and you could drop this.

But I suppose that doesn't matter all that much.

opudrovs replied:
Can you please post sample code?

opudrovs replied:
Replaced.

ozamosi commented on ui/lib/utils.ts:
```diff
@@ -161,3 +169,122 @@ export const convertImage = (image: string) => {
+    : (automation as HelmRelease)?.helmChart?.sourceRef;
+}
+
+// findNode searches for a node by the node's name and namespace.
```
Is it worth just putting the nodes into a map, and then we wouldn't need to search?

opudrovs replied:
Done.

ozamosi commented on ui/lib/objects.ts:
```diff
@@ -199,3 +200,48 @@ export class Provider extends FluxObject {
+  constructor(fluxObject: FluxObject, isCurrentNode?: boolean) {
+    this.obj = fluxObject.obj;
+    this.uid = fluxObject.uid;
+    this.displayKind = removeKind(fluxObject.kind);
```
This can just be `fluxObject.type`

opudrovs replied:
OK, good.

opudrovs replied:
Done.

ozamosi commented on ui/hooks/objects.ts:
```diff
@@ -66,3 +74,29 @@ export function useGetObject<T extends FluxObject>(
+
+type Res = { objects: FluxObject[]; errors: ListError[] };
+
+export function useListObjects<T extends FluxObject>(
```
Didn't we say something on… Friday was it? About how there can be objects in different clusters with the same name, so we need to make sure this only queries the right cluster?

opudrovs replied:
AH, so it was not implemented yet? OK, I will look into it.

ozamosi replied:
I had a sudden realization: now that this feature is here, we will be able to use it in `SourceDetails` for populating the relevant Automation list.

ozamosi commented on ui/lib/dependencies.ts:
```diff
@@ -0,0 +1,111 @@
+
+  currentNode.isCurrentNode = true;
+
+  // Find nodes in the current node's dependency tree.
+  let graphNodes: FluxObjectNode[] = [];
+
+  const visitedNodes = new Map<string, boolean>();
+
+  let nodesToExplore: FluxObjectNode[] = [currentNode].concat(
+    getNeighborNodes(nodes, currentNode)
+  );
+
+  if (nodesToExplore.length === 1) {
+    return [];
+  }
+
+  while (nodesToExplore.length > 0) {
+    const node = nodesToExplore.shift();
+
+    const newNodes = getNeighborNodes(nodes, node).filter(
+      (n) => !visitedNodes[n.id]
+    );
+
+    for (const n of newNodes) {
+      visitedNodes[n.id] = true;
+    }
+
+    nodesToExplore = nodesToExplore.concat(newNodes);
+
+    graphNodes = graphNodes.concat(node);
+  }
+
+  // Remove duplicates from graphNodes.
+  graphNodes = graphNodes.filter(
+    (node, index) => graphNodes.indexOf(node) === index
+  );
```
This is what I meant about moving bits around to avoid having to remove duplicates:

```suggestion
  // Find nodes in the current node's dependency tree.
  let graphNodes: FluxObjectNode[] = [];
  
  const visitedNodes = new Map<string, boolean>();
  visitedNodes[currentNode.id] = true;

  let nodesToExplore: FluxObjectNode[] = [currentNode];
  
  while (nodesToExplore.length > 0) {
    const node = nodesToExplore.shift();
    
    const newNodes = getNeighborNodes(nodes, node).filter(
      (n) => !visitedNodes[n.id]
    );
    
    for (const n of newNodes) {
      visitedNodes[n.id] = true;
    }
    
    nodesToExplore = nodesToExplore.concat(newNodes);
    
    graphNodes = graphNodes.concat(node);
  }
  
  if (graphNodes.length === 1) {
    return [];
  }
```

opudrovs commented:
@ozamosi @joshri

- I reworked the links/edges a bit because with the same lines as in the old graph it looked like all three kustomizations in the second row in the bigger dependencies graph depend on kustomization 1 while only kustomization e should depend on kustomizationn 1. I hope now it is more informative.

<img width="1508" alt="Screenshot 2022-09-14 at 05 05 46" src="https://user-images.githubusercontent.com/8824708/190050793-267b8136-8b5b-4110-a6c8-94ab61ac1424.png">

<img width="1525" alt="Screenshot 2022-09-14 at 05 06 12" src="https://user-images.githubusercontent.com/8824708/190050833-7b96dfb3-622e-409c-9d55-b8305f252993.png">

I can improve how the edges and arrows are drawn further but we'll need to decide if then should be shaped as triangles or arrow tips + if lines should go under non-right angles with arrows rotated based on line direction or if lines should be curved if straight angles should be preserved for arrows (similar to the way they are displayed right now) and so on.

- Also made the message a reusable component:

<img width="1517" alt="Screenshot 2022-09-14 at 05 08 34" src="https://user-images.githubusercontent.com/8824708/190051856-4a2f82a1-a9fe-4115-a422-224487bd9e54.png">

- For now I combined the DAG and Directed graph nodes, it was rather easy. The graphs stay separate. Directed graph nodes now support isCurrentNode prop (I've added it for the `secondNode`).

opudrovs commented:
Other comments were addressed too, I hope I haven't missed anything.

ozamosi reviewed (approved):
:tada:
